src/main/scala/org/constellation/snapshotstreaming/GlobalSnapshotReader.scala

```diff
@@ -0,0 +1,64 @@
+package org.constellation.snapshotstreaming
+
+import cats.data.OptionT
+import cats.effect.Async
+import cats.syntax.applicativeError._
+import cats.syntax.either._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.option._
+
+import org.tessellation.schema.{GlobalSnapshot, IncrementalGlobalSnapshot}
+import org.tessellation.sdk.domain.cluster.storage.L0ClusterStorage
+import org.tessellation.sdk.http.p2p.clients.L0GlobalSnapshotClient
+import org.tessellation.security.Hashed
+import org.tessellation.security.hash.Hash
+import org.tessellation.security.signature.Signed
+
+import io.circe.Codec
+import io.circe.generic.semiauto.deriveCodec
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+trait GlobalSnapshotReader[F[_]] {
+
+  def readGlobalSnapshot(hash: Hash): F[Either[Signed[GlobalSnapshot], Signed[IncrementalGlobalSnapshot]]]
+}
+
+object GlobalSnapshotReader {
+
+  def make[F[_]: Async](
+    globalSnapshotClient: L0GlobalSnapshotClient[F],
+    l0ClusterStorage: L0ClusterStorage[F],
+    snapshotStorage: FileBasedLastGlobalSnapshotStorage[F]
+  ): GlobalSnapshotReader[F] = new GlobalSnapshotReader[F] {
+    private val logger = Slf4jLogger.getLogger[F]
+
+    implicit val codec: Codec[Hashed[GlobalSnapshot]] = deriveCodec[Hashed[GlobalSnapshot]]
```

Comment:
We can extract it if it's reused in multiple places in snapshot-streaming to not redefine it in separate classes. The thing about no codec for `Hashed` in tessellation is that it's not supposed to be sent over network, it should be for internal usage only. No supplied codec kind of helps for it to be that way.